File: src/main/java/core/procedures/interop/PrimitiveNumberType.java
package core.procedures.interop;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;

public final class PrimitiveNumberType extends AFn {

  public static final PrimitiveNumberType BYTE   = new PrimitiveNumberType(byte.class);
  public static final PrimitiveNumberType SHORT  = new PrimitiveNumberType(short.class);
  public static final PrimitiveNumberType INT    = new PrimitiveNumberType(int.class);
  public static final PrimitiveNumberType LONG   = new PrimitiveNumberType(long.class);
  public static final PrimitiveNumberType DOUBLE = new PrimitiveNumberType(double.class);
  public static final PrimitiveNumberType FLOAT  = new PrimitiveNumberType(float.class);

  private final String name;
  private final Class clazz;

  private PrimitiveNumberType(Class clazz) {
    super(new FnArgsBuilder().minArgs(1).maxArgs(1).mandatoryArgsTypes(new Class[]{Number.class}));
    this.name = clazz.getSimpleName();
    this.clazz = clazz;
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return name;
  }

  /* FIXME Have to box it */
  @Override
  public Number apply1(Object arg) {
    Number number = (Number)arg;
    if (clazz == byte.class) {
      return number.byteValue();
    } else if (clazz == short.class) {
      return number.shortValue();
    } else if (clazz == int.class) {
      return number.intValue();
    } else if (clazz == long.class) {
      return number.longValue();
    } else if (clazz == double.class) {
      return number.doubleValue();
    } else if (clazz == float.class) {
      return number.floatValue();
    }
    throw new IllegalArgumentException("Unknown primitive type: " + clazz);
  }
}

File: src/main/java/core/procedures/interop/PrimitiveNumberType.kt
package core.procedures.interop

import core.procedures.AFn
import core.procedures.Arity.Exactly

open class PrimitiveNumberType private constructor(override val name: String, private val function: (Number) -> Number) :
        AFn<Number?, Number>(arity = Exactly(1), isPure = true, mandatoryArgsTypes = arrayOf(Number::class.java)) {

    companion object {
        val BYTE   = PrimitiveNumberType("byte",   Number::toByte)
        val SHORT  = PrimitiveNumberType("short",  Number::toShort)
        val INT    = PrimitiveNumberType("int",    Number::toInt)
        val LONG   = PrimitiveNumberType("long",   Number::toLong)
        val FLOAT  = PrimitiveNumberType("float",  Number::toFloat)
        val DOUBLE = PrimitiveNumberType("double", Number::toDouble)
    }

    override operator fun invoke(arg: Number?) = function(arg!!)
}
